### PR TITLE
Fixed #215

### LINKED ISSUE #215: Allow setting write mode in configuration file
If someone needs a specific write mode (for example overwrite, if one is using version control and does not need to keep backup files), they should be able to specify it in the configuration once, and then just run `rustfmt` without always having to specify `--write-mode`.

### PR DESCRIPTION
**Pull request by svmnotn**
The code is a hack.
That is the best way of putting it, but it works.

I did think of maybe re-writing how WriteMode works as it seems, to me at lest, that not all the features are implemented/documented. That or I might have understood all of it, I'm pretty new at Rust.

### CODE REVIEW

**Comment by marcusklaas** [comment]
Thanks for the PR! Do I understand correctly that there are now two enums for the write mode? We'd have just one, ideally. Also, it's no longer possible to set the write mode to `Replace` with command line parameters once the write mode is specified in the config, right? I think the best way to solve this is to have a distinct `Default` variant, which maps to `<WriteMode as Default>::default()`.

**Comment by svmnotn** [comment]
Firstly, no problem since this is something I want for my projects as well.

To answer some of your questions:
* Yes there are 2 enums, which is why I referred to it as a hack. I would also like just one but the parsing code was a bit much for me as it would need a change so that it could support the `NewFile(&'static str)` option in lib.rs' WriteMode.
* Yes, I did so because when the write-mode param is not passed in, `Replace` is what is assumed.
* Where would we put Default? In lib.rs' WriteMode?

**Comment by marcusklaas** [comment]
I'd say we drop the `NewFile` variant. I can't think of any use for it and doubt any one is using it. It doesn't work well when multiple files are involved either. @nrc, what do you think?

In that case we should be able to re-use `WriteMode` in `lib.rs`. We'd just need an invocation of `impl_enum_decodable` to be able to parse it.

**Comment by svmnotn** [comment]
If we are going to drop variants, we should take the opportunity and check that they are all needed so that we only need to make this change once.

**Comment by marcusklaas** [comment]
The other ones all have their use, I think. `WriteMode::Return` is mostly useful when rustfmt is used as a library and not so much as a binary. To remove it would require some non-trivial refactoring, so I'd leave it for now.

**Comment by svmnotn** [comment]
I was looking at how to go about removing the second enum and found that WriteMode implements `FromStr`, which is also implemented in `impl_enum_decodable!`, however the current implementation returns an error when `Return` is converted. Is this intended?

**Comment by marcusklaas** [comment]
Yea, because it shouldn't ever be used by users of the binary.

**Comment by svmnotn** [comment]
How can this function be maintained across the `impl_enum_decodable!` call?

**Comment by marcusklaas** [comment]
It cannot. I guess we'll have to remove `Return` as a variant after all. Sorry for the confusion.

**Comment by svmnotn** [comment]
Ok, no problem. Right now what I did with default is to make a variant so that if both the config and the argument are default it will use Replace but if there is a argument it will prefer that, however in filemap.rs there is a match against WriteMode, how should I implement the case where Default is the mode? Which should never happen in the execution of the program, a `panic!`?

**Comment by marcusklaas** [comment]
You can use `unreachable!()` there. If, by some bug, it would be evaluated in runtime the program will panic. It signals to the reader that we *should* never go there.

**Comment by svmnotn** [comment]
Got it, that is what I ended up writing since it also ends up as a panic. Should I put those changes as a new commit or rebase the current one?

**Comment by marcusklaas** [comment]
A rebase would be cleanest, I think.

**Comment by svmnotn** [comment]
Should I keep WriteMode in lib.rs or move it to config since at this point using `configuration_option_enum` results in the same code?

**Comment by marcusklaas** [comment]
Both make sense to me. Either way is fine!

**Comment by svmnotn** [comment]
Opps... Forgot to change tests...

File: src/config.rs [hunk 1]
```diff
@@ -113,6 +113,24 @@ configuration_option_enum! { ReportTactic:
     Never,
 }
 
+configuration_option_enum! { WriteMode:
+  // Used internally to represent when no option is given
```
Comment (marcusklaas) [inline src/config.rs]: Nit: could you use 4 space indents here?

**Comment by svmnotn** [comment]
Ok so many of the tests use the Return option, not quite sure how to handle this, any ideas?

File: src/config.rs (hunk 1 above)
Comment (svmnotn) [inline src/config.rs]: I guess rustfmt doesn't change the indents...

Comment (marcusklaas) [inline src/config.rs]: Not inside these macro thingies (yet).

**Comment by marcusklaas** [comment]
Yea, that's the rough bit. The basic idea I had was to have a public function that returns a `HashMap` from file names (`String`s) to formatted code (`String`s). The tests would use that.

The binary could use that function to either write the results to stdout or file, depending on `WriteMode`.

This is actually a change I've been wanting to do for a while, so I wouldn't mind finishing up your PR (by making a PR for your branch). If you're willing to give it a try, that'd be even cooler. Let me know.

**Comment by svmnotn** [comment]
I would be fine with the branch idea, at the end all I want is for the feature to be there.

However I did think of doing a check on the run function so that it panics if anyone uses return there, not sure if that is a good way around it though, however the tests never run correctly on my machine(Windows 8), even when they pass in Travis.

**Comment by svmnotn** [comment]
That function also sound like a nice abstraction on the `write_file` fn so if you want to use the opportunity to do that go ahead and create the branch and I'll reopen to that branch

**Comment by marcusklaas** [comment]
Ah, that may be https://github.com/rust-lang-nursery/rustfmt/issues/617. I'm not sure how to fix that as I'm not on a Windows machine myself..

I'll try to make a PR to your branch tonight. Thanks for your changes!

**Comment by svmnotn** [comment]
No problem! Thank you very much for your comments

**Comment by marcusklaas** [comment]
I'll just create a PR to your master branch ;-)

**Comment by svmnotn** [comment]
Just a note, I just realized that `Default` is a trait... I did not now that, however the code works better having an implicit `Default` variant since it gets rid of the ambiguity of `Replace`

**Comment by marcusklaas** [comment]
Agreed!

**Comment by marcusklaas** [comment]
Created a pull request for your branch here: https://github.com/svmnotn/rustfmt/pull/1.

**Comment by marcusklaas** [comment]
Hmmm, one unintended consequence seems to be that write modes are now case sensitive. That's probably not what we want.

**Comment by svmnotn** [comment]
yeah... I ran into that while testing but I think so are line endings and everything using the derived fromStr (I think, I need to test that)

**Comment by svmnotn** [comment]
Just tested, line endings are also case sensitive

**Comment by marcusklaas** [comment]
The line endings option, you mean?

**Comment by svmnotn** [comment]
yes, sorry I was looking at the code, I'll branch off and see if maybe I can PR that in. Hopefully it will be less problematic ;)

**Comment by svmnotn** [comment]
#739 fixes the issue pointed out by @marcusklaas

**Comment by nrc** [comment]
@marcusklaas is this landable now?

@svmnotn needs a rebase

**Pull request closed without merging**